File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public boolean helper(TreeNode root, long min, long max){
        if(root == null) return true;

        if(root.val>min && root.val<max){
            return helper(root.left,min,root.val) && helper(root.right,root.val,max);
        }
        return false;
    }


    public boolean isValidBST(TreeNode root) {
        return helper(root,Long.MIN_VALUE, Long.MAX_VALUE);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self,root,min_val,max_val):
        if(root == None): return True

        if(min_val<root.val<max_val):
            return self.helper(root.left,min_val,root.val) and self.helper(root.right,root.val,max_val)
        else:
            return False    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.helper(root, float('-inf'), float('inf'))

